### PR TITLE
Check indexable data address correct on object movement

### PR DESCRIPTION
**Pull request by jonoommen**
Extend `gcCheck` to ensure new `dataAddr` field for indexable objects
is correctly initialized and updated on indexable object movement
for both contiguous and discontiguous arrays.

Related to: https://github.com/eclipse-openj9/openj9/issues/11438

Signed-off-by: Jonathan Oommen <jon.oommen@gmail.com>

### CODE REVIEW

**Comment by dmitripivkine** [comment]
Please check your DDR change handles properly:
- old cores (old Indexable object header)
- transitional cores (new Indexable object header but not initialized or updated dataAddr - NULL, stall address or even garbage)
- new cores (obviously)

**Comment by jonoommen** [comment]
@amicic should the DDR changes be introduced in this PR, or should it be introduced later in a later phase of https://github.com/eclipse-openj9/openj9/issues/11438?

File: runtime/gc_check/CheckEngine.cpp [hunk 1]
```diff
@@ -460,6 +460,15 @@ GC_CheckEngine::checkJ9Object(J9JavaVM *javaVM, J9Object* objectPtr, J9MM_Iterat
 		}
 	}
 
+#if defined(J9VM_ARCH_X86) && defined(J9VM_ENV_DATA64)
```
Comment (amicic) [inline runtime/gc_check/CheckEngine.cpp:463]: I would not prevent users to enable this check on P&Z platforms (both runtime or for core checks). You should even try and verify it does work and reports warnings at one of the 2. 

From what I can see, this check is not _on_ by default, yet, what is ok, for now.

**Comment by jonoommen** [comment]
It was requested that I ask for a review from both of you @dmitripivkine and @keithc-ca. Would you both be able to review this?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/J9IndexableObjectHelper.java [hunk 2]
```diff
@@ -142,6 +142,36 @@ public static long discontiguousHeaderSize()
 		return J9IndexableObjectDiscontiguous.SIZEOF;
 	}
 
+	/**
+	 * @param objPointer the contiguous array object who's dataAddr field we are accessing
+	 * @throws CorruptDataException
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/J9IndexableObjectHelper.java:147]: No trailing whitespace please (on any modified or added line).

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/J9IndexableObjectHelper.java [hunk 3]
```diff
@@ -142,6 +142,36 @@ public static long discontiguousHeaderSize()
 		return J9IndexableObjectDiscontiguous.SIZEOF;
 	}
 
+	/**
+	 * @param objPointer the contiguous array object who's dataAddr field we are accessing
+	 * @throws CorruptDataException 
+	 */
+	public static VoidPointer getDataAddrForContiguous(J9IndexableObjectPointer objPointer) throws CorruptDataException, NoSuchFieldException
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/J9IndexableObjectHelper.java:149]: You should document the possibility of `NoSuchFieldException` (although I'm still deciding whether it should be handled here).

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java [hunk 4]
```diff
@@ -79,7 +79,8 @@
 	public static final int J9MODRON_GCCHK_MISC_ALWAYS_DUMP_STACK = 0x00004000;
 	public static final int J9MODRON_GCCHK_MISC_DARKMATTER = 0x00008000;
 	public static final int J9MODRON_GCCHK_MISC_MIDSCAVENGE = 0x00010000;
-	public static final int J9MODRON_GCCHK_MISC_OWNABLESYNCHRONIZER_CONSISTENCY = 0x00020000;
+	public static final int J9MODRON_GCCHK_VALID_INDEXABLE_DATA_ADDRESS = 0x00020000;
+	public static final int J9MODRON_GCCHK_MISC_OWNABLESYNCHRONIZER_CONSISTENCY = 0x00040000;
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java:83]: None of the constants `J9MODRON_GCCHK_*` should be defined here. They should be accessed from DDR-generated code derived from `CheckBase.hpp` to avoid the duplication and potential disagreement.

The value of `J9MODRON_GCCHK_MISC_OWNABLESYNCHRONIZER_CONSISTENCY` is now wrong for old core files. These will need to be made available to DDR and listed in `CompatibilityConstants29.dat`.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckCycle.java [hunk 5]
```diff
@@ -343,6 +343,11 @@ private void initialize(String options)
 							miscFlags |= J9MODRON_GCCHK_MISC_MIDSCAVENGE;
 							continue;
 						}
+
+						if(miscOption.equals("indexabledataaddress")) {
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckCycle.java:347]: formatting: space after `if`

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java [hunk 6]
```diff
@@ -1189,6 +1194,29 @@ private int checkJ9Object(J9ObjectPointer object, GCHeapRegionDescriptor regionD
...
+				if (J9IndexableObjectHelper.isInlineContiguousArraylet(array)) {
+					isCorrectData = J9IndexableObjectHelper.getDataAddrForContiguous(array).equals(VoidPointer.cast(array.addOffset(J9IndexableObjectHelper.contiguousHeaderSize())));
+				} else {
+					isCorrectData = J9IndexableObjectHelper.getDataAddrForDiscontiguous(array).equals(VoidPointer.cast(array.addOffset(J9IndexableObjectHelper.discontiguousHeaderSize())));
+				}
+
+				if (false == isCorrectData) {
+					return J9MODRON_GCCHK_RC_INVALID_INDEXABLE_DATA_ADDRESS;
+				}
+			} catch (NoSuchFieldException e) {
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java:1211]: The body of the `try` block should be reduced the only the part that might throw `NoSuchFieldException`.

File: runtime/gc_check/CheckBase.hpp [hunk 7]
```diff
@@ -156,6 +156,9 @@
 #define J9MODRON_GCCHK_RC_REPLACED_CLASS_HAS_NO_HOTSWAP_FLAG ((UDATA) 40)
 #define J9MODRON_GCCHK_RC_CLASS_IS_UNLOADED ((UDATA) 48)
 
+/* Special error codes for when J9MODRON_GCCHK_VALID_INDEXABLE_DATA_ADDRESS is set */
+#define J9MODRON_GCCHK_RC_INVALID_INDEXABLE_DATA_ADDRESS ((UDATA) 34)
+
```
Comment (keithc-ca) [inline runtime/gc_check/CheckBase.hpp:161]: Please move this up so the definitions stay in numerical order.

File: runtime/gc_check/CheckBase.hpp [hunk 8]
```diff
@@ -101,6 +101,7 @@
 #define J9MODRON_GCCHK_MISC_ALWAYS_DUMP_STACK ((UDATA)0x00004000)
 #define J9MODRON_GCCHK_MISC_DARKMATTER ((UDATA)0x00008000)
 #define J9MODRON_GCCHK_MISC_MIDSCAVENGE ((UDATA)0x00010000)
+#define J9MODRON_GCCHK_VALID_INDEXABLE_DATA_ADDRESS ((UDATA)0x00020000)
```
Comment (keithc-ca) [inline runtime/gc_check/CheckBase.hpp:104]: These definitions should be exposed to DDR to avoid repeating the definitions in java code.

File: runtime/gc_check/CheckEngine.cpp [hunk 9]
```diff
@@ -460,6 +460,15 @@ GC_CheckEngine::checkJ9Object(J9JavaVM *javaVM, J9Object* objectPtr, J9MM_Iterat
 		}
 	}
 
+#if defined(J9VM_ENV_DATA64)
+	if ((extensions->isVLHGC() && _cycle->getMiscFlags() & J9MODRON_GCCHK_VALID_INDEXABLE_DATA_ADDRESS) && extensions->objectModel.isIndexable(objectPtr)) {
```
Comment (keithc-ca) [inline runtime/gc_check/CheckEngine.cpp:464]: This should use `OMR_ARE_ANY_BITS_SET`:
```
... && OMR_ARE_ANY_BITS_SET(_cycle->getMiscFlags(), J9MODRON_GCCHK_VALID_INDEXABLE_DATA_ADDRESS) && ...
```

File: runtime/gc_check/CheckReporterTTY.cpp [hunk 10]
```diff
@@ -196,15 +200,16 @@ GC_CheckReporterTTY::report(GC_CheckError *error)
...
 GC_CheckReporterTTY::reportObjectHeader(GC_CheckError *error, J9Object *objectPtr, const char *prefix)
 {
+	UDATA headerSize;
 	MM_GCExtensionsBase* extensions = MM_GCExtensions::getExtensions(_javaVM);
 	const char *prefixString = prefix ? prefix : "";
-	UDATA headerSize = extensions->objectModel.getHeaderSize(objectPtr);
 	PORT_ACCESS_FROM_PORT(_portLibrary);
 
-	if (!shouldReport(error)) {
+	if (!shouldReport(error) || NULL == J9GC_J9OBJECT_CLAZZ(objectPtr, extensions)) {
```
Comment (keithc-ca) [inline runtime/gc_check/CheckReporterTTY.cpp:208]: Parentheses please for `==` subexpression.

File: runtime/gc_glue_java/ArrayletObjectModel.hpp [hunk 11]
```diff
@@ -931,6 +931,30 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
...
+	/**
+	 * Asserts that the dataAddr field of the indexable object is correct.
+	 *
+	 * @param arrayPtr      Pointer to the indexable object
+	 * @param dataAddr      Pointer to indexable object data address
+	 */
+	MMINLINE bool
+	isCorrectDataAddrContiguous(J9IndexableObject *arrayPtr, void *dataAddr)
+	{
+		return (dataAddr == ((void *)((uintptr_t)arrayPtr + contiguousHeaderSize())));
```
Comment (keithc-ca) [inline runtime/gc_glue_java/ArrayletObjectModel.hpp:943]: No parentheses for return expression; also line 955 below.

File: runtime/gc_glue_java/ArrayletObjectModel.hpp [hunk 12]
```diff
@@ -991,6 +1015,46 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
...
+	 */
+	MMINLINE bool
+	isCorrectDataAddr(J9IndexableObject *arrayPtr)
+	{
+		ArrayLayout layout = getArrayLayout(arrayPtr);
+		bool isDiscontiguous = (layout != InlineContiguous);
+
+		return isDiscontiguous
+			? isCorrectDataAddrForDiscontiguousArraylet(arrayPtr)
+			: isCorrectDataAddrForContiguousArraylet(arrayPtr);
```
Comment (keithc-ca) [inline runtime/gc_glue_java/ArrayletObjectModel.hpp:1032]: Why the reverse logic? Why not:
```
	return (InlineContiguous == layout)
		? isCorrectDataAddrForContiguousArraylet(arrayPtr)
		: isCorrectDataAddrForDiscontiguousArraylet(arrayPtr);
```

File: runtime/gcchk/module.xml [hunk 13]
```diff
@@ -78,6 +78,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 			<library name="omrgcstructs" type="external"/>
 			<library name="j9gcstats"/>
 			<library name="omrgcstats" type="external"/>
+			<library name="omrglue" type="external"/>
```
Comment (keithc-ca) [inline runtime/gcchk/module.xml:81]: Why is this needed? Why not also for cmake builds?

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/gc_glue_java/ArrayletObjectModel.hpp (hunk 12 above)
Comment (jonoommen) [reply runtime/gc_glue_java/ArrayletObjectModel.hpp:1032]: I used the formatting of the method above for consistency, even though I didn't like it. I fully agree this logic is better than the reverse logic and will gladly change it.

File: runtime/gcchk/module.xml (hunk 13 above)
Comment (jonoommen) [reply runtime/gcchk/module.xml:81]: It was needed to fix a link failure I was having with my Aaxon builds on Windows. This is the link failure that was fixed from this change: 

`error LNK2019: unresolved external symbol:
"public: enum GC_ArrayletObjectModelBase::ArrayLayout __cdecl GC_ArrayletObjectModel::getArrayletLayout(struct J9Class *,unsigned __int64,unsigned __int64)" (?getArrayletLayout@GC_ArrayletObjectModel@@QEAA?AW4ArrayLayout@GC_ArrayletObjectModelBase@@PEAUJ9Class@@_K1@Z)
referenced in function:
"private: unsigned __int64 __cdecl GC_CheckEngine::checkJ9Object(struct J9JavaVM *,struct J9Object *,struct J9MM_IterateRegionDescriptor *,unsigned __int64)" (?checkJ9Object@GC_CheckEngine@@AEAA_KPEAUJ9JavaVM@@PEAUJ9Object@@PEAUJ9MM_IterateRegionDescriptor@@_K@Z)`

Comment (keithc-ca) [reply runtime/gcchk/module.xml:81]: What about the other half of my question: why not also for cmake?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/J9IndexableObjectHelper.java (hunk 2 above)
Comment (dmitripivkine) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/J9IndexableObjectHelper.java:147]: should comment mentioned throws NoSuchFieldException as well? - here and next function

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java [hunk 14]
```diff
@@ -1189,6 +1194,29 @@ private int checkJ9Object(J9ObjectPointer object, GCHeapRegionDescriptor regionD
 				return ret;
 			}
 		}
+
+		if (GCExtensions.isVLHGC() && J9BuildFlags.env_data64 && isIndexableDataAddressFlagSet() && ObjectModel.isIndexable(object)) {
```
Comment (dmitripivkine) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java:1198]: Is there a reason why check is limited to Balanced only? I know it supposed to be used in Balanced for now but is something prevents is to do this check for Standard collectors? We should update dataAddr properly too. Metronome never moves objects, so initial value should be correct also. I would prefer not to check for GC policy here if it is possible.

**Comment by dmitripivkine** [comment]
There is number of resolved conversations but I don't see correspondent changes. Is something is not committed yet?

File: runtime/gcchk/module.xml (hunk 13 above)
Comment (jonoommen) [reply runtime/gcchk/module.xml:81]: It was not required for CMake builds (not entirely sure why, but CMake doesn't have any issue linking `ArrayletObjectModelBase` in `CheckEngine.cpp`

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java (hunk 4 above)
Comment (dmitripivkine) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java:83]: I believe this is an exceptional case when constants are used for internal GC Check purpose only (run time flag to request check feature but not other JVM functionality). They might be or not to be the same between DDR GC Check vs. Runtime GC Check. Also I think there is no backwards compatibility issue for `J9MODRON_GCCHK_MISC_OWNABLESYNCHRONIZER_CONSISTENCY` as far as it is used run time only (Am I missing something?)
Currently we keep constants for DDR GC Check and Runtime GC Check synchronized but there is no functional dependency between them.
So if in general I agree with statement we should not allow duplication for definitions in this particular case I am not sure we want to create dependency between DDR and Runtime GC Checks
@amicic What do you think?

Comment (jonoommen) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java:83]: Ok I see. Should I make this change for all `J9MODRON_GCCHK_*` constants and update DDR code for all these constants accordingly?

Comment (keithc-ca) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java:83]: As long as those constants don't represent anything that might exist in a core file, they don't need to be stable (or even match the native code values). That the names are the same makes the answer unclear.

Comment (amicic) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java:83]: Since they are not part of core files and don't impose compatibility issue, I would not add new changes to sync them in this PR. But, in general, it would be nice they are synced, if someone finds time/will to do the work.

Comment (dmitripivkine) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java:83]: >Ok I see. Should I make this change for all J9MODRON_GCCHK_* constants and update DDR code for all these constants accordingly?

I think this place is ok "as is" and no further changes required for this PR. If we decided to do cleaning/renaming/refactoring it can be addressed in different request

Comment (keithc-ca) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java:83]: On the other hand, there's no need to change the values used in native code, so the existing names would remain in sync.

File: runtime/gc_check/CheckBase.hpp (hunk 8 above)
Comment (jonoommen) [reply runtime/gc_check/CheckBase.hpp:104]: I believe this is now resolved based on the discussion and comment https://github.com/eclipse-openj9/openj9/pull/12910/files#r676847957

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java (hunk 14 above)
Comment (jonoommen) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java:1198]: I don't believe there should be anything preventing us from doing this check in standard collectors. I will test this to make sure.

File: runtime/gc_check/CheckBase.hpp (hunk 8 above)
Comment (keithc-ca) [reply runtime/gc_check/CheckBase.hpp:104]: I don't agree. I'll resolve this and make new comments where I think they're needed.

File: runtime/gc_check/CheckReporterTTY.cpp (hunk 10 above)
Comment (keithc-ca) [reply runtime/gc_check/CheckReporterTTY.cpp:208]: This has not been fixed.

Comment (dmitripivkine) [reply runtime/gc_check/CheckReporterTTY.cpp:208]: I am curious why this change is required. What is discovered as an object with NULL class in it (and why it was not required before)?

File: runtime/gc_glue_java/ArrayletObjectModel.hpp (hunk 11 above)
Comment (keithc-ca) [reply runtime/gc_glue_java/ArrayletObjectModel.hpp:943]: This is not fixed.

File: runtime/gc_glue_java/ArrayletObjectModel.hpp (hunk 12 above)
Comment (keithc-ca) [reply runtime/gc_glue_java/ArrayletObjectModel.hpp:1032]: Are you going to change it as suggested?

File: runtime/gcchk/module.xml (hunk 13 above)
Comment (keithc-ca) [reply runtime/gcchk/module.xml:81]: I don't see the `omrglue` library in the cmake world. @dnakamura Any theories why not?

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java [hunk 15]
```diff
@@ -124,6 +125,9 @@
 
 	public static final int J9MODRON_GCCHK_RC_OWNABLE_SYNCHRONIZER_INVALID_CLASS = 38;
 
+	/* Special error codes for when J9MODRON_GCCHK_VALID_INDEXABLE_DATA_ADDRESS is set */
+	public static final int J9MODRON_GCCHK_RC_INVALID_INDEXABLE_DATA_ADDRESS = 34;
```
[the remark this answers is not in the document]
Comment (dmitripivkine) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java:129]: I believe constants here are grouped based on functionality not in order - class check related, dead object related etc. I think this is correct that new group is created - dataAddr related checks

File: runtime/gc_check/CheckReporterTTY.cpp (hunk 10 above)
Comment (jonoommen) [reply runtime/gc_check/CheckReporterTTY.cpp:208]: It was a corner case that came up when testing a particular benchmark (I think it was the Hazelcast Jet Streaming benchmark). I can't recall the details exactly as I debugged this weeks ago, but I am sure this was required to handle the particular corner case.

Comment (dmitripivkine) [reply runtime/gc_check/CheckReporterTTY.cpp:208]: I think we should look closer to this case (not necessarily in this PR). In theory it should not happen. We should figure out why unexpected happened - is it defect of GC Check itself or there is a larger problem behind

Comment (dmitripivkine) [reply runtime/gc_check/CheckReporterTTY.cpp]: If it is possible to reproduce the problem and generate system core for it (using `abort` sub option for example) I can help to figure out what happened.

**Comment by jonoommen** [comment]
> There is number of resolved conversations but I don't see correspondent changes. Is something is not committed yet?

Yes, that's correct, I had resolved them locally but was waiting for a discussion with one of the comments to wrap-up before pushing my changes. They are committed now 👍

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (jonoommen) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckBase.java]: I have made some minor changes to both CheckBase.hpp and CheckBase.java to make them both more consistent with each other.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java [hunk 16]
```diff
@@ -1189,6 +1194,30 @@ private int checkJ9Object(J9ObjectPointer object, GCHeapRegionDescriptor regionD
...
+				}
+			} catch (NoSuchFieldException e) {
+				/*
+				 * Do nothing - NoSuchFieldException from trying to access the indexable object field "dataAddr"
+				 * is due to the incorrect usage of gccheck misc option "indexabledataaddress"
+				 * on a core file that was generated from a build where the "dataAddr" field does not exists yet
+				 */
+				_cycle.clearIndexableDataAddrCheckMiscFlag();
+			}
+
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java:1216]: No trailing whitespace, please.

File: runtime/gc_check/CheckReporterTTY.cpp [hunk 17]
```diff
@@ -196,15 +200,16 @@ GC_CheckReporterTTY::report(GC_CheckError *error)
 void 
 GC_CheckReporterTTY::reportObjectHeader(GC_CheckError *error, J9Object *objectPtr, const char *prefix)
 {
+	UDATA headerSize;
```
Comment (keithc-ca) [inline runtime/gc_check/CheckReporterTTY.cpp:203]: All scalar local variables should be initialized in the declaration.

File: runtime/gc_glue_java/ArrayletObjectModel.cpp [hunk 18]
```diff
@@ -63,13 +63,13 @@ GC_ArrayletObjectModel::AssertArrayletIsDiscontiguous(J9IndexableObject *objPtr)
...
 {
-        Assert_MM_true(InlineContiguous == getArrayLayout(objPtr));
+	Assert_MM_true(InlineContiguous == getArrayLayout(objPtr));
 }
 
 void
 GC_ArrayletObjectModel::AssertDiscontiguousArrayletLayout(J9IndexableObject *objPtr)
 {
-        Assert_MM_true(Discontiguous == getArrayLayout(objPtr));
+	Assert_MM_true(InlineContiguous != getArrayLayout(objPtr));
```
Comment (keithc-ca) [inline runtime/gc_glue_java/ArrayletObjectModel.cpp:72]: Why this change? I think the old condition was more appropriate.

File: runtime/gc_glue_java/ArrayletObjectModel.hpp [hunk 19]
```diff
@@ -931,6 +931,30 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
...
+	/**
+	 * Asserts that the dataAddr field of the indexable object is correct.
+	 *
+	 * @param arrayPtr      Pointer to the indexable object
+	 * @param dataAddr      Pointer to indexable object data address
+	 */
+	MMINLINE bool
+	isCorrectDataAddrContiguous(J9IndexableObject *arrayPtr, void *dataAddr)
+	{
+		return dataAddr == ((void *)((uintptr_t)arrayPtr + contiguousHeaderSize()));
```
Comment (keithc-ca) [inline runtime/gc_glue_java/ArrayletObjectModel.hpp:943]: Redundant parentheses, it could be:
```
  return dataAddr == (void *)((uintptr_t)arrayPtr + contiguousHeaderSize());
```
Same comment for line 955.

File: runtime/gc_glue_java/ArrayletObjectModel.hpp [hunk 20]
```diff
@@ -931,6 +931,30 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
...
+	}
+
```
Comment (keithc-ca) [inline runtime/gc_glue_java/ArrayletObjectModel.hpp:945]: No trailing whitespace, please.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/gc_glue_java/ArrayletObjectModel.cpp (hunk 18 above)
Comment (jonoommen) [reply runtime/gc_glue_java/ArrayletObjectModel.cpp:72]: Yes the old condition was more clear, however ever it does not account for `getArrayLayout` returning `Hybrid`, which should not trigger this assertion. Alternatively, perhaps this would be better:

`ArrayLayout layout = getArrayLayout(objPtr);`
`Assert_MM_true(Discontiguous == layout || (Hybrid == layout));`

File: runtime/gc_glue_java/ArrayletObjectModel.hpp (hunk 20 above)
Comment (jonoommen) [reply runtime/gc_glue_java/ArrayletObjectModel.hpp:945]: My mistake. Thought I checked all lines for trailing white space but apparently missed checking empty lines.

File: runtime/gc_glue_java/ArrayletObjectModel.cpp (hunk 18 above)
Comment (jonoommen) [reply runtime/gc_glue_java/ArrayletObjectModel.cpp:72]: Yes, the old condition was more appropriate, however, it was incorrect as it did not account for array layout `Hydrid` which should not trigger the assert. However, perhaps this is best:

`ArrayLayout layout = getArrayLayout(objPtr);`
`Assert_MM_true((Discontiguous == layout) || (Hybrid == layout));`

**Comment by keithc-ca** [comment]
keith line endings check

(testing a possible solution to #13246).

**Comment by keithc-ca** [comment]
```
13:43:41  + git diff --check origin/master HEAD
13:43:41  debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ObjectModel.java:465: trailing whitespace.
13:43:41  +	 * @throws CorruptDataException 
13:43:41  debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCObjectModel.java:301: trailing whitespace.
13:43:41  +	 * @throws CorruptDataException 
13:43:41  debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/J9IndexableObjectHelper.java:191: trailing whitespace.
13:43:41  +	 * @throws CorruptDataException 
13:43:41  debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/gccheck/CheckEngine.java:1216: trailing whitespace.
13:43:41  +			
13:43:41  runtime/gc_glue_java/ArrayletObjectModel.hpp:945: trailing whitespace.
13:43:41  +	
```

**Comment by jonoommen** [comment]
keith line endings check

**Comment by keithc-ca** [comment]
keith line endings check

**Review by keithc-ca — APPROVED** [review]

File: runtime/gc_check/CheckEngine.cpp [hunk 21]
```diff
@@ -460,6 +460,15 @@ GC_CheckEngine::checkJ9Object(J9JavaVM *javaVM, J9Object* objectPtr, J9MM_Iterat
 		}
 	}
 
+#if defined(J9VM_ENV_DATA64)
+	if (OMR_ARE_ANY_BITS_SET(_cycle->getMiscFlags(), J9MODRON_GCCHK_VALID_INDEXABLE_DATA_ADDRESS) && extensions->objectModel.isIndexable(objectPtr)) {
+		/* Check that the indexable object has the correct data address pointer */
+		if (false == extensions->indexableObjectModel.isCorrectDataAddr((J9IndexableObject*)objectPtr)) {
```
Comment (dmitripivkine) [inline runtime/gc_check/CheckEngine.cpp:466]:
```suggestion
		if (!extensions->indexableObjectModel.isCorrectDataAddr((J9IndexableObject*)objectPtr)) {
```

File: runtime/gc_check/CheckReporterTTY.cpp [hunk 22]
```diff
@@ -198,13 +202,13 @@ GC_CheckReporterTTY::reportObjectHeader(GC_CheckError *error, J9Object *objectPt
 {
 	MM_GCExtensionsBase* extensions = MM_GCExtensions::getExtensions(_javaVM);
 	const char *prefixString = prefix ? prefix : "";
-	UDATA headerSize = extensions->objectModel.getHeaderSize(objectPtr);
 	PORT_ACCESS_FROM_PORT(_portLibrary);
 
-	if (!shouldReport(error)) {
+	if (!shouldReport(error) || (NULL == J9GC_J9OBJECT_CLAZZ(objectPtr, extensions))) {
```
Comment (dmitripivkine) [inline runtime/gc_check/CheckReporterTTY.cpp:207]: Would you please open side item to investigate why it is required?

**Review by dmitripivkine — APPROVED** [review]
LGTM with minor requests

**Review by amicic — APPROVED** [review]

**Comment by amicic** [comment]
Jenkins test sanity win,aix jdk11
